src/repository.c

```diff
@@ -2477,26 +2477,78 @@ int git_repository_state_cleanup(git_repository *repo)
 	return git_repository__cleanup_files(repo, state_files, ARRAY_SIZE(state_files));
 }
 
-int git_repository_is_shallow(git_repository *repo)
+int git_repository__shallow_roots(git_array_oid_t *out, git_repository *repo)
 {
 	git_buf path = GIT_BUF_INIT;
+	git_buf contents = GIT_BUF_INIT;
 	struct stat st;
-	int error;
+	int error, updated;
+	size_t start, end;
 
 	if ((error = git_buf_joinpath(&path, repo->path_repository, "shallow")) < 0)
 		return error;
 
 	error = git_path_lstat(path.ptr, &st);
-	git_buf_free(&path);
 
 	if (error == GIT_ENOTFOUND) {
+		git_buf_free(&path);
 		giterr_clear();
 		return 0;
 	}
 
+	if (out == NULL) {
+		git_buf_free(&path);
+		return st.st_size == 0 ? 0 : 1;
+	}
+
+	error = git_futils_readbuffer_updated(&contents, git_buf_cstr(&path), &repo->shallow_checksum, &updated);
+	git_buf_free(&path);
+
 	if (error < 0)
 		return error;
-	return st.st_size == 0 ? 0 : 1;
+
+	if (!updated) {
+		goto unchanged;
+	}
+
+	git_array_clear(repo->shallow_oids);
+
+	start = end = 0;
+	while (end < contents.size) {
+		if (contents.ptr[end] == '\n' || end == contents.size) {
+			if (end - start == GIT_OID_HEXSZ) {
+				git_oid *oid = git_array_alloc(repo->shallow_oids);
+				error = git_oid_fromstrn(oid, contents.ptr + start, GIT_OID_HEXSZ);
+				if (error < 0) {
+					/* TODO: Warn ? */
```

Comment:
Yeah, that feels better actually. Also, I don't like that loop's nesting, I'll try to cleanup. If you have some idea, I'll take it.